### PR TITLE
[Python] Add Python commissioning flow

### PR DESCRIPTION
**Pull request by erjiaqing**
https://github.com/project-chip/connectedhomeip/issues/24055

This PR introduces manually commissioning flow in Python, with a example of commissioning flow without autocommissioner. This is useful for testing potential commissioning flow changes in the future since it is more flexible. It can also be used to test a few corner cases during commissioning, including bad certificate chain etc.

This PR also splits the Controller in Python into two parts, ControllerBase (like Controller in C++, actually backed with a Commissioner with dummy CertificateIssuer), and Controller (the Commissioner), the ControllerBase can be initialized with a certificate signed by someone else.

This PR only includes changes to Python scripts and necessary C++ interfaces.

The changes to CHIP Tool will be done later in separate PR.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

File: src/controller/python/OpCredsBinding.cpp [hunk 1]
```diff
@@ -369,11 +373,74 @@ void pychip_OnCommissioningStatusUpdate(chip::PeerId peerId, chip::Controller::C
...
+ *
+ */
+PyChipError pychip_OpCreds_AllocateControllerForPythonCommissioningFLow(chip::Controller::DeviceCommissioner ** outDevCtrl,
+                                                                        chip::python::pychip_P256Keypair * operationalKey,
+                                                                        uint8_t * noc, uint32_t nocLen, uint8_t * icac,
+                                                                        uint32_t icacLen, uint8_t * rcac, uint32_t rcacLen,
+                                                                        const uint8_t * ipk, uint32_t ipkLen,
+                                                                        chip::VendorId adminVendorId, bool enableServerInteractions)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
Comment (andy31415) [inline src/controller/python/OpCredsBinding.cpp:388]: nit: move this where first used? Smaller scope makes things easier to read

File: src/controller/python/OpCredsBinding.cpp [hunk 2]
```diff
@@ -369,11 +373,74 @@ void pychip_OnCommissioningStatusUpdate(chip::PeerId peerId, chip::Controller::C
     return sTestCommissioner.OnCommissioningStatusUpdate(peerId, stageCompleted, err);
 }
 
+/**
+ * Call of pychip_OpCreds_AllocateControllerFoCustomCommissioningFlow allows allocating a device commissioner that will perform
+ * commissioning without using the auto-commissioning flow present in the Controller class
+ *
+ */
+PyChipError pychip_OpCreds_AllocateControllerForPythonCommissioningFLow(chip::Controller::DeviceCommissioner ** outDevCtrl,
```
Comment (andy31415) [inline src/controller/python/OpCredsBinding.cpp:381]: These methods take a lot of arguments and some seem to be representable by byte arrays (i.e. they are data + length pointers). 

Could we pass in a @dataclass containing the information needed and not have a method with 9 parameters?

Could be a followup TODO, however anything with more than 4 arguments is very hard to maintain.

File: src/controller/python/OpCredsBinding.cpp [hunk 3]
```diff
@@ -369,11 +373,74 @@ void pychip_OnCommissioningStatusUpdate(chip::PeerId peerId, chip::Controller::C
     return sTestCommissioner.OnCommissioningStatusUpdate(peerId, stageCompleted, err);
 }
 
+/**
+ * Call of pychip_OpCreds_AllocateControllerFoCustomCommissioningFlow allows allocating a device commissioner that will perform
```
Comment (andy31415) [inline src/controller/python/OpCredsBinding.cpp:377]: nit: missing a `r` from the name of the method. Maybe don't put the name there at all as it is very long anyway. Just say `Allocates a controller that does not use auto-commisioning`.

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 4]
```diff
@@ -1467,3 +1313,222 @@ def _InitLib(self):
 
             self._dmLib.pychip_DeviceController_GetLogFilter = [None]
             self._dmLib.pychip_DeviceController_GetLogFilter = c_uint8
+
+            self._dmLib.pychip_OpCreds_AllocateController.argtypes = [c_void_p, POINTER(
+                c_void_p), c_uint64, c_uint64, c_uint16, c_char_p, c_bool, c_bool, POINTER(c_uint32), c_uint32, c_void_p]
+            self._dmLib.pychip_OpCreds_AllocateController.restype = PyChipError
+
+            self._dmLib.pychip_OpCreds_AllocateControllerForPythonCommissioningFLow.argtypes = [
+                POINTER(c_void_p), c_void_p, c_char_p, c_uint32, c_char_p, c_uint32, c_char_p, c_uint32, c_char_p, c_uint32, c_uint16, c_bool]
```
Comment (andy31415) [inline src/controller/python/chip/ChipDeviceCtrl.py:1322]: c_char_p in https://docs.python.org/3/library/ctypes.html is documented as `(NUL terminated)`. should we just say c_void_p? Not sure if it matters though.

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 5]
```diff
@@ -1467,3 +1313,222 @@ def _InitLib(self):
...
+
+
+class ChipDeviceController(ChipDeviceControllerBase):
+    ''' The ChipDeviceCommissioner binding, named as ChipDeviceController
+
+    TODO: This class contains DEPRECATED functions, we should update the test scripts to avoid the usage of those functions.
+    '''
+
+    def __init__(self, opCredsContext: ctypes.c_void_p, fabricId: int, nodeId: int, adminVendorId: int, catTags: typing.List[int] = [], paaTrustStorePath: str = "", useTestCommissioner: bool = False, fabricAdmin: FabricAdmin = None, name: str = None, keypair: p256keypair.P256Keypair = None):
+        super().__init__("caIndex(%x)/fabricId(0x%016X)/nodeId(0x%016X)" % (fabricAdmin.caIndex, fabricId, nodeId) if name is None else name)
```
Comment (andy31415) [inline src/controller/python/chip/ChipDeviceCtrl.py:1333]: nit: would 

```py
super().__init___(
    name or 
    f"caIndex({fabricAdmin.caIndex:x})/fabricId(0x{fabridIc:%016X})/nodeId(0x{nodeId:%016X})"
)
``` 

be more readable ?

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 6]
```diff
@@ -1467,3 +1313,222 @@ def _InitLib(self):
...
+        self.CheckIsActive()
+        self._ChipStack.commissioningCompleteEvent.clear()
+        self.state = DCState.COMMISSIONING
+
+        self._ChipStack.CallAsync(
+            lambda: self._dmLib.pychip_DeviceController_Commission(
+                self.devCtrl, nodeid)
+        )
+        if not self._ChipStack.commissioningCompleteEvent.isSet():
+            # Error 50 is a timeout
```
Comment (andy31415) [inline src/controller/python/chip/ChipDeviceCtrl.py:1380]: This comment confuses me: what is `Error 50`? Is this a copy & paste?

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 7]
```diff
@@ -1467,3 +1313,222 @@ def _InitLib(self):
...
+
+    @property
+    def caIndex(self) -> int:
+        return self._caIndex
+
+    @property
+    def fabricAdmin(self) -> FabricAdmin:
+        return self._fabricAdmin
+
+    def Commission(self, nodeid):
```
Comment (andy31415) [inline src/controller/python/chip/ChipDeviceCtrl.py:1370]: A doc comment about returning a bool may help (or at least type it as `-> bool`

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 6 above)
Comment (andy31415) [reply src/controller/python/chip/ChipDeviceCtrl.py:1380]: This looks like a:

```py
return (self._ChipStack.commissioningCompleteEvent.isSet() and 
       (self._ChipStack.commissioningEventRes == 0))
```

File: src/controller/python/chip/commissioning/pase.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        if self.is_ble:
+            self.devCtrl.CloseBLEConnection(self.is_ble)
+
+
+def establish_session(devCtrl: ChipDeviceCtrl.ChipDeviceControllerBase, parameter: commissioning.PaseParameters) -> ContextManager:
+    if isinstance(parameter, commissioning.PaseOverBLEParameters):
+        devCtrl.EstablishPASESessionBLE(parameter.setup_pin, parameter.discriminator, parameter.temporary_nodeid)
+    elif isinstance(parameter, commissioning.PaseOverIPParameters):
+        device = devCtrl.DiscoverCommissionableNodes(filterType=discovery.FilterType.LONG_DISCRIMINATOR,
+                                                     filter=parameter.discriminator, stopOnFirst=True)
```
Comment (andy31415) [inline src/controller/python/chip/commissioning/pase.py:52]: it seems like we want to rename `discriminator` to `long_discriminator` since this is the filter we are using.

File: src/controller/python/chip/commissioning/pase.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        if not device:
+            raise ValueError("No commissionable device found")
+        selected_address = None
+        for ip in device[0].addresses:
+            if ipaddress.ip_address(ip).is_link_local:
+                # TODO(erjiaqing): The link local address in the discovery result does not contain a interface
```
Comment (andy31415) [inline src/controller/python/chip/commissioning/pase.py:58]: What does this mean? don't they work? is this a specific environment or a bug in the python implementation? We should be able to use link local addresses to commission and not devices have non-link-local addresses.

You could write a sort to prefer non-link local addresses if you want, however skipping them completely seems odd/wrong.

File: src/controller/python/chip/crypto/p256keypair.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    VerifyOrReturnError(mDeriveSecretFunct(mPyContext, remote_public_key.ConstBytes(), out_secret.Bytes(), &secretLength),
+                        CHIP_ERROR_INTERNAL);
+    out_secret.SetLength(secretLength);
+
+    return CHIP_NO_ERROR;
+}
+
+void pychip_P256Keypair::UpdatePubkey(uint8_t * publicKey)
+{
+    memcpy(mPublicKey.Bytes(), publicKey, mPublicKey.Length());
```
Comment (andy31415) [inline src/controller/python/chip/crypto/p256keypair.cpp:69]: how do we know both key sizes are identical? could we pass in a span instead?

File: src/controller/python/chip/crypto/p256keypair.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+    mInitialized = true;
+}
+
+chip::python::pychip_P256Keypair * pychip_NewP256Keypair(void * pyObject, pychip_P256Keypair_ECDSA_sign_msg aSignMsgFunct,
+                                                         pychip_P256Keypair_ECDH_derive_secret aDeriveSecretFunct)
+{
+    auto res = new pychip_P256Keypair(pyObject, aSignMsgFunct, aDeriveSecretFunct);
```
Comment (andy31415) [inline src/controller/python/chip/crypto/p256keypair.cpp:77]: what method does one call to `delete` this memory allocation? 

I believe a comment regarding expected lifetime would help... I see no pychip_DeleteP256Keypair method.

File: src/controller/python/chip/crypto/p256keypair.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+pychip_P256Keypair::pychip_P256Keypair(void * aPyContext, pychip_P256Keypair_ECDSA_sign_msg aSignMsgFunct,
+                                       pychip_P256Keypair_ECDH_derive_secret aDeriveSecretFunct) :
+    mPyContext(aPyContext),
+    mSignMsgFunct(aSignMsgFunct), mDeriveSecretFunct(aDeriveSecretFunct)
+{}
+
+pychip_P256Keypair::~pychip_P256Keypair() {}
+
+CHIP_ERROR pychip_P256Keypair::Initialize(Crypto::ECPKeyTarget key_target)
```
Comment (andy31415) [inline src/controller/python/chip/crypto/p256keypair.cpp:19]: if this is empty, why do we have it?

File: src/controller/python/chip/crypto/p256keypair.h [hunk 13]
```diff
@@ -0,0 +1,105 @@
+#pragma once
```
Comment (andy31415) [inline src/controller/python/chip/crypto/p256keypair.h:1]: add license blurb to files?

File: src/controller/python/chip/crypto/p256keypair.cpp [hunk 14]
```diff
@@ -0,0 +1,85 @@
+
```
Comment (andy31415) [inline src/controller/python/chip/crypto/p256keypair.cpp:1]: license blurb?

**Comment by andy31415** [comment]
@erjiaqing - flake8 seems to complain about things. If code is correct we can add an exception, otherwise we need to fix things.

File: src/controller/python/OpCredsBinding.cpp (hunk 1 above)
Comment (erjiaqing) [reply src/controller/python/OpCredsBinding.cpp]: Done

File: src/controller/python/OpCredsBinding.cpp (hunk 2 above)
Comment (erjiaqing) [reply src/controller/python/OpCredsBinding.cpp:382]: https://github.com/project-chip/connectedhomeip/issues/25214

Already working on this.

File: src/controller/python/OpCredsBinding.cpp (hunk 3 above)
Comment (erjiaqing) [reply src/controller/python/OpCredsBinding.cpp]: done

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 5 above)
Comment (erjiaqing) [reply src/controller/python/chip/ChipDeviceCtrl.py]: done

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 6 above)
Comment (erjiaqing) [reply src/controller/python/chip/ChipDeviceCtrl.py]: Done

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 7 above)
Comment (erjiaqing) [reply src/controller/python/chip/ChipDeviceCtrl.py]: done

File: src/controller/python/chip/crypto/p256keypair.h (hunk 13 above)
Comment (erjiaqing) [reply src/controller/python/chip/crypto/p256keypair.h:19]: done

File: src/controller/python/chip/crypto/p256keypair.cpp (hunk 14 above)
Comment (erjiaqing) [reply src/controller/python/chip/crypto/p256keypair.cpp:18]: done

File: src/controller/python/chip/crypto/p256keypair.cpp (hunk 12 above)
Comment (erjiaqing) [reply src/controller/python/chip/crypto/p256keypair.cpp:40]: To override the default Initialize() function from the platform.

File: src/controller/python/chip/crypto/p256keypair.cpp (hunk 11 above)
Comment (erjiaqing) [reply src/controller/python/chip/crypto/p256keypair.cpp:97]: Added pychip_DeleteP256Keypair

File: src/controller/python/chip/crypto/p256keypair.cpp (hunk 10 above)
Comment (erjiaqing) [reply src/controller/python/chip/crypto/p256keypair.cpp]: Updated and added a check

File: src/controller/python/chip/commissioning/pase.py (hunk 9 above)
Comment (erjiaqing) [reply src/controller/python/chip/commissioning/pase.py]: Updated the comment.

Althouth we can use the ll address for commissioning, however, the address retrieved by the API does not contain the interface identification part, and some function may fail.

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 4 above)
Comment (erjiaqing) [reply src/controller/python/chip/ChipDeviceCtrl.py]: Updated to POINTER(c_char), (but `POINTER(c_uint8_t)` won't work)

**Comment by erjiaqing** [comment]
> @erjiaqing - flake8 seems to complain about things. If code is correct we can add an exception, otherwise we need to fix things.

Updated the script to follow flake8's suggestions.
